### PR TITLE
Refine InferShape for recurrent_network_op

### PR DESCRIPTION
**Pull request by qingqing01**
The tensor only contains the shape and does not hold memory when inferring shape.

### CODE REVIEW

File: paddle/operators/recurrent_network_op.cc [hunk 1]
```diff
@@ -42,28 +43,33 @@ void SegmentInputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
...
       step_input->Resize(step_dims);
     }
   }
 }
 
 void ConcatOutputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
                    const std::vector<Link>& outlinks,
-                   const size_t seq_len) {
+                   const size_t seq_len,
+                   bool infer_shape) {
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op.cc]: is_infer or infer_mode?

File: paddle/operators/recurrent_network_op.cc [hunk 2]
```diff
@@ -97,18 +104,14 @@ void LinkMemories(std::vector<std::shared_ptr<Scope>>& scopes,
   std::shared_ptr<Scope> scope = scopes[step_id];
   std::shared_ptr<Scope> linked_scope = scopes[step_id + offset];
   for (auto& attr : memories) {
-    auto mem = scope->CreateVariable(attr.pre_var)->GetMutable<Tensor>();
+    auto mem = scope->GetVariable(attr.pre_var)->GetMutable<Tensor>();
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op.cc]: need an enforce on `scope->GetVariable(xxx) != nullptr` ?

File: paddle/operators/recurrent_network_op.cc [hunk 3]
```diff
@@ -176,61 +179,43 @@ void RecurrentAlgorithm::InferShape(const std::shared_ptr<Scope>& scope) const {
...
                  "stepnet [%s] is not in scope.",
                  arg_->step_net);
   Variable* net = scope->GetVariable(arg_->step_net);
   PADDLE_ENFORCE(net != nullptr, "failed to get step net");
-  // If the InferShape is called in OperatorBase's run function,
-  // the rnn op only needs to do InferShape for the first time step
   for (size_t i = 0; i < seq_len_; i++) {
     if (i > 0) {
-      rnn::LinkMemories(step_scopes, arg_->memories, i, -1);
+      rnn::LinkMemories(step_scopes, arg_->memories, i, -1, true);
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op.cc]: `true /* infer_mode */`

File: paddle/operators/recurrent_network_op.cc [hunk 4]
```diff
@@ -176,61 +179,43 @@ void RecurrentAlgorithm::InferShape(const std::shared_ptr<Scope>& scope) const {
...
-                         ->GetMutable<Tensor>()
-                         ->dims();
-    std::vector<int> dims_vec = vectorize(step_dims);
-    // now only support fixed length
-    dims_vec.insert(dims_vec.begin(), seq_len_);
-    Tensor* output =
-        step_scopes[0]->GetVariable(outlinks[i].external)->GetMutable<Tensor>();
-    output->Resize(make_ddim(dims_vec));
-  }
+  rnn::ConcatOutputs(step_scopes, arg_->outlinks, seq_len_, true);
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op.cc]: same as top

File: paddle/operators/recurrent_network_op.cc [hunk 5]
```diff
@@ -176,61 +179,43 @@ void RecurrentAlgorithm::InferShape(const std::shared_ptr<Scope>& scope) const {
...
+
+  InitMemories(step_scopes[0], false);
+
   Variable* net = scope->GetVariable(arg_->step_net);
   for (size_t step_id = 0; step_id < seq_len_; step_id++) {
-    // the link memory is done in InferShape
-    // maybe remove following code after testing
     if (step_id > 0) {
-      rnn::LinkMemories(step_scopes, arg_->memories, step_id, -1);
+      rnn::LinkMemories(step_scopes, arg_->memories, step_id, -1, false);
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op.cc]: same as the top

File: paddle/operators/recurrent_network_op_test.cc [hunk 6]
```diff
@@ -152,14 +150,11 @@ class RecurrentOpTest : public ::testing::Test {
...
     auto net = var->GetMutable<NetOp>();
-    // rnn/s is net's input or output?
-    net->inputs_ = {"rnn/h@pre", "rnn/w", "rnn/x"};
-    net->inputs_ = {"rnn/s", "rnn/h"};
     net->AddOp(
         OpRegistry::CreateOp("mul", {"rnn/h@pre", "rnn/w"}, {"rnn/s"}, {}));
 
     net->AddOp(
-        OpRegistry::CreateOp("add_two", {"rnn/x", "rnn/s"}, {"rnn/h"}, {}));
+        OpRegistry::CreateOp("add_two", {"x@alias", "rnn/s"}, {"rnn/h"}, {}));
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op_test.cc]: an inline function for the action of adding `@alias` suffix ?

File: paddle/operators/recurrent_network_op_test.cc [hunk 7]
```diff
@@ -318,7 +313,7 @@ class RecurrentGradientAlgorithmTest : public ::testing::Test {
         scope_->GetVariable("step_scopes")
             ->GetMutable<std::vector<std::shared_ptr<Scope>>>();
     for (int i = 1; i < 10; ++i) {
-      rnn::LinkMemories(*step_scopes, memories, i, -1);
+      rnn::LinkMemories(*step_scopes, memories, i, -1, true);
```
Comment (Superjom) [inline paddle/operators/recurrent_network_op_test.cc]: same as the top, `true /* infer_mode */

**Comment by luotao1** [comment]
recurrentOp里可以调用infershape，也可以不调用infershape?请问分别是什么时候呢

File: paddle/operators/recurrent_network_op_test.cc [hunk 8]
```diff
@@ -347,7 +342,7 @@ TEST(RecurrentOp, LinkMemories) {
     scope->CreateVariable("pre_h");
     auto tensor = scope->CreateVariable("h")->GetMutable<Tensor>();
     float* data = tensor->mutable_data<float>(make_ddim({15, 20}), CPUPlace());
-    for (int i = 0; i < 15 * 20; ++i) {
+    for (int j = 0; j < 15 * 20; ++j) {
       data[i] = rand() * (1. / (double)RAND_MAX);
```
Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op_test.cc]: data[i]   ==> data[j]  ?

File: paddle/operators/recurrent_network_op_test.cc [hunk 9]
```diff
@@ -347,7 +342,7 @@ TEST(RecurrentOp, LinkMemories) {
     scope->CreateVariable("pre_h");
     auto tensor = scope->CreateVariable("h")->GetMutable<Tensor>();
     float* data = tensor->mutable_data<float>(make_ddim({15, 20}), CPUPlace());
-    for (int i = 0; i < 15 * 20; ++i) {
+    for (int j = 0; j < 15 * 20; ++j) {
```
Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op_test.cc]: data[i] ==> data[j] ?

Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op_test.cc]: data[i] ==> data[j] ?

File: paddle/operators/recurrent_network_op.cc [hunk 10]
```diff
@@ -29,7 +29,8 @@ namespace rnn {
 
 void SegmentInputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
                    const std::vector<Link>& inlinks,
-                   const size_t seq_len) {
+                   const size_t seq_len,
```
Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op.cc]: const size_t&

File: paddle/operators/recurrent_network_op.cc [hunk 11]
```diff
@@ -42,28 +43,33 @@ void SegmentInputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
...
+      }
       step_input->Resize(step_dims);
     }
   }
 }
 
 void ConcatOutputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
                    const std::vector<Link>& outlinks,
-                   const size_t seq_len) {
+                   const size_t seq_len,
```
Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op.cc]: const size_t&

File: paddle/operators/recurrent_network_op.cc [hunk 12]
```diff
@@ -79,8 +85,9 @@ void ConcatOutputs(std::vector<std::shared_ptr<Scope>>& step_scopes,
 
 void LinkMemories(std::vector<std::shared_ptr<Scope>>& scopes,
                   const std::vector<rnn::MemoryAttr>& memories,
-                  size_t step_id,
-                  int offset) {
+                  const size_t step_id,
```
Comment (jacquesqiao) [inline paddle/operators/recurrent_network_op.cc]: const size_t&

File: paddle/operators/recurrent_network_op.cc (hunk 1 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: infer_shape_mode

File: paddle/operators/recurrent_network_op.cc (hunk 2 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: Done.

File: paddle/operators/recurrent_network_op.cc (hunk 3 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: Done.

File: paddle/operators/recurrent_network_op.cc (hunk 4 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: Done.

File: paddle/operators/recurrent_network_op.cc (hunk 5 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: Done.

File: paddle/operators/recurrent_network_op_test.cc (hunk 6 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op_test.cc]: This unit test may move to Python, so not add an inline function for adding @alias suffix.

File: paddle/operators/recurrent_network_op_test.cc (hunk 7 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op_test.cc]: Done.

File: paddle/operators/recurrent_network_op_test.cc (hunk 9 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op_test.cc]: Done

File: paddle/operators/recurrent_network_op.cc (hunk 10 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: `const size_t` is ok.

File: paddle/operators/recurrent_network_op.cc (hunk 11 above)
Comment (qingqing01) [inline paddle/operators/recurrent_network_op.cc]: `const size_t` is ok.

File: paddle/operators/recurrent_op.cc [hunk 13]
```diff
@@ -336,35 +323,37 @@ void RecurrentGradientAlgorithm::Run(
...
     if (static_cast<size_t>(step_id) != seq_len_ - 1) {
-      rnn::LinkMemories(step_scopes, arg_->memories, step_id, 1);
+      rnn::LinkMemories(
+          step_scopes, arg_->memories, step_id, 1, false /*infer_shape_mode*/);
     }
     net->GetMutable<NetOp>()->Run(step_scopes[step_id], dev_ctx);
   }
-  LinkBootMemoryGradients(step_scopes[0]);
-  rnn::ConcatOutputs(step_scopes, arg_->outlinks, seq_len_);
+  LinkBootMemoryGradients(step_scopes[0], false);
```
Comment (Superjom) [inline paddle/operators/recurrent_op.cc]: false add a comment ?

File: paddle/operators/recurrent_op_test.cc [hunk 14]
```diff
@@ -362,7 +361,7 @@ TEST(RecurrentOp, LinkMemories) {
   memories.push_back(mem_attr);
 
   for (int i = 1; i < len; ++i) {
-    rnn::LinkMemories(step_scopes, memories, i, -1);
+    rnn::LinkMemories(step_scopes, memories, i, -1, false /*infer_shape_mode*/);
   }
   // check
   for (int i = 0; i < len - 1; ++i) {
```
Comment (Superjom) [inline paddle/operators/recurrent_op_test.cc]: unify all the array index from `int` to `size_t`

**Pull request merged**